### DISCUSSION TITLE
Building RustDesk to produce executables like the official releases doesn't work

### DISCUSSION THREAD

**Discussion by davidramiro** [rustdesk/rustdesk#695] [General]
First of all, thanks a lot for this project. 

**Describe the bug you encountered:**

---

Building as per tutorial won't produce release files as expected. (Yes, I have sciter.dll file present in build directory)

Building with `cargo run` runs the app fine. Building with `cargo run --release` builds an executable file, that will not work on its own. It will show a blank screen. I have found out, that you can fix this by manually putting `ui` folder into build folder.

I have looked into `build.py` and the docs [on here](https://rustdesk.com/docs/en/dev/build/windows/). I tried the following: 

Created `resources` folder, put in `src/ui` and `sciter.dll`. Used `cargo build --features with_rc`. Executable works, but it shows a console window in addition.

How to build a single executable that will just show the app, no console, got everything linked? I searched through all the issues, but no definitive answer.

Are there any build commands I am missing?

---

**How did you install `RustDesk`?**

Built from source. Tried latest `master` and `1.1.9` tag.

---

**Environment**

Same behaviour on Arch Linux and Windows 10. I have yet to test mac OS, will do later.

**Comment by rustdesk** [comment]
https://github.com/rustdesk/rustdesk/issues/501#issuecomment-1125596473

**Reply by davidramiro** [reply]
I saw that issue. This is not my problem here. I am able to work with the dynamic sciter dll without a problem. But the program behaves differently than on release executable. Showing a console window. How to hide?

**Reply by rustdesk** [reply]
Everything is in build.py
https://github.com/rustdesk/rustdesk/blob/5182b96bda96119bcdd21d8be09e7451cfa0f339/build.py#L118

**Reply by davidramiro** [reply]
Thanks! That comment is what I needed! So switching from `cargo build` to `python3 build.py` already was a big help. Sciter files are now linked.

Now just one more small question: When I install RustDesk after clicking on UAC notice, how can I make RustDesk copy over the sciter.dll to Program Files as well? Is there a function ready? If not, can you point me to the function where it gets installed into Program Files?

**Reply by rustdesk** [reply]
> how can I make RustDesk copy over the sciter.dll to Program Files as well? 

No, we have no this, because we bought Sciter static lib.

**Reply by davidramiro** [reply]
No problem, I can add that to the code myself, but where exactly do the files get written to the installation directory? I suppose it's on `platform/windows.rs`, or am I wrong? Which line for copying files to output directory?

**Comment by youhers** [comment]
Thank you for your method, it works thank you

**Comment by youhers** [comment]
> 首先，非常感谢这个项目。
> 
> **描述你遇到的错误：**
> 
> 按照教程构建不会按预期生成发布文件。（是的，我在构建目录中有 sciter.dll 文件）
> 
> 构建与`cargo run`运行应用程序正常。使用`cargo run --release`构建构建一个可执行文件，该文件将无法单独运行。它将显示一个空白屏幕。我发现，您可以通过手动将`ui`文件夹放入构建文件夹来解决此问题。
> 
> 我已经查看了[这里](https://rustdesk.com/docs/en/dev/build/windows/)`build.py`的文档。我尝试了以下方法：[](https://rustdesk.com/docs/en/dev/build/windows/)
> 
> 创建`resources`文件夹，放入`src/ui`和`sciter.dll`. 用过`cargo build --features with_rc`。可执行文件有效，但它还显示了一个控制台窗口。
> 
> 如何构建一个只显示应用程序、没有控制台、链接所有内容的可执行文件？我搜索了所有问题，但没有明确的答案。
> 
> 我缺少任何构建命令吗？
> 
> **你是怎么安装的`RustDesk`？**
> 
> 从源代码构建。尝试了最新`master`的和`1.1.9`标记。
> 
> **环境**
> 
> Arch Linux 和 Windows 10 上的行为相同。我还没有测试 mac OS，稍后再做。

Can I ask you, I will follow your method

**Reply by davidramiro** [reply]
I added 

```
let src_exe = std::env::current_exe()?;
let cur_dir = src_exe.parent().unwrap();
let sciter_dll = cur_dir.join("sciter.dll").to_str().unwrap_or("").to_owned();
```

to the top of the `install_me()` function in `src/platform/windows.rs` to copy the path of sciter.dll in the executable directory. Then i put 

```
copy /Y \"{sciter_dll}\" \"{path}\"
```

into the script on [line 1025](https://github.com/rustdesk/rustdesk/blob/master/src/platform/windows.rs#L1025).

Then I build the app with `python3 build.py`, which calls `python3 inline-sciter.py` in line 121. This links everything together. Now you can run the app with `sciter.dll` in the same directory (or buy license from Sciter to link statically). When you install the service on Windows, `sciter.dll` now gets copied to the program files folder.

**Reply by rtest12** [reply]
> into the script on [line 1025](https://github.com/rustdesk/rustdesk/blob/master/src/platform/windows.rs#L1025).
> 
> Then I build the app with `python3 build.py`, which calls `python3 inline-sciter.py` in line 121. This links everything together. Now you can run the app with `sciter.dll` in the same directory (or buy license from Sciter to link statically). When you install the service on Windows, `sciter.dll` now gets copied to the program files folder.

Could you please share in more detail about how you did it, step by step? I followed your instructions, now UI folder is working, but without scyter.dll still does not start, tell me, what else did I miss?

**Reply by davidramiro** [reply]
> Could you please share in more detail about how you did it, step by step? I followed your instructions, now UI folder is working, but without scyter.dll still does not start, tell me, what else did I miss?

The built exe will still need sciter.dll to be present. Just when you install the windows service, sciter.dll will get copied to the destination path. This is all that that code does.

**Comment by yashpatel1599** [comment]
@davidramiro @youhers have you try to build .deb linux executable ?
